homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -117,7 +118,7 @@ def domain(self) -> str:
     @property
     def redirect_uri(self) -> str:
         """Return the redirect uri."""
-        return f"{self.hass.config.api.base_url}{AUTH_CALLBACK_PATH}"  # type: ignore
+        return f"{async_get_url(self.hass)}{AUTH_CALLBACK_PATH}"
```

Comment:
@balloob For the `redirect_uri` of an OAuth2 conf flow, I wonder if we should prefer to use the external URL here?

Reply:
Well. To be honest there is no need whatsoever for it to be external. Most oauth2 provider even allow localhost. It just need to be resolvable from the browser you are doing the flow on.

So it's just a matter of it matching what user configured it as on the other end.

Reply:
Yeah, in that case, the current will do: "Give me anything".

Reply:
Yes. hairpin is a potential problem here. But I think, if external url is configured, use that, otherwise local. Not sure what you mean by cloud url here. This code would be the non cloud version of this helper. So it's external or internal.

For cloud it's a completely different redirect endpoint.

Reply:
"Get anything" is about problematic, since you need to know in advance which url the system will use when you configure the remote service (some support multiple redirect URLs btw). 

we need to be able to write in our documentation which url it will use.

Reply:
> Not sure what you mean by cloud url here. This code would be the non cloud version of this helper. So it's external or internal.

It is the URL of the remote UI of the Home Assistant Cloud if the user has enabled that. The nice thing about using that method, is that it works for users with and without hairpinning problems. However, it is indeed not ideal either.

So let stick internal if configured, else external, worst case if everything else fails, use the remote UI (which is the default which is in right now).